src/udev/cdrom_id/cdrom_id.c

```diff
@@ -722,581 +438,544 @@ static int cd_profiles(int fd) {
-        if (cd_media_dvd_r_dl)
-                printf("ID_CDROM_MEDIA_DVD_R_DL=1\n");
-        if (cd_media_dvd_r_dl_seq)
-                printf("ID_CDROM_MEDIA_DVD_R_DL_SEQ=1\n");
-        if (cd_media_dvd_r_dl_jr)
-                printf("ID_CDROM_MEDIA_DVD_R_DL_JR=1\n");
-        if (cd_media_dvd_ram)
-                printf("ID_CDROM_MEDIA_DVD_RAM=1\n");
-        if (cd_media_dvd_rw)
-                printf("ID_CDROM_MEDIA_DVD_RW=1\n");
-        if (cd_media_dvd_rw_dl)
-                printf("ID_CDROM_MEDIA_DVD_RW_DL=1\n");
-        if (cd_media_dvd_plus_r)
-                printf("ID_CDROM_MEDIA_DVD_PLUS_R=1\n");
-        if (cd_media_dvd_plus_rw)
-                printf("ID_CDROM_MEDIA_DVD_PLUS_RW=1\n");
-        if (cd_media_dvd_plus_rw_dl)
-                printf("ID_CDROM_MEDIA_DVD_PLUS_RW_DL=1\n");
-        if (cd_media_dvd_plus_r_dl)
-                printf("ID_CDROM_MEDIA_DVD_PLUS_R_DL=1\n");
-        if (cd_media_bd)
-                printf("ID_CDROM_MEDIA_BD=1\n");
-        if (cd_media_bd_r)
-                printf("ID_CDROM_MEDIA_BD_R=1\n");
-        if (cd_media_bd_r_srm)
-                printf("ID_CDROM_MEDIA_BD_R_SRM=1\n");
-        if (cd_media_bd_r_rrm)
-                printf("ID_CDROM_MEDIA_BD_R_RRM=1\n");
-        if (cd_media_bd_re)
-                printf("ID_CDROM_MEDIA_BD_RE=1\n");
-        if (cd_media_hddvd)
-                printf("ID_CDROM_MEDIA_HDDVD=1\n");
-        if (cd_media_hddvd_r)
-                printf("ID_CDROM_MEDIA_HDDVD_R=1\n");
-        if (cd_media_hddvd_r_dl)
-                printf("ID_CDROM_MEDIA_HDDVD_R_DL=1\n");
-        if (cd_media_hddvd_ram)
-                printf("ID_CDROM_MEDIA_HDDVD_RAM=1\n");
-        if (cd_media_hddvd_rw)
-                printf("ID_CDROM_MEDIA_HDDVD_RW=1\n");
-        if (cd_media_hddvd_rw_dl)
-                printf("ID_CDROM_MEDIA_HDDVD_RW_DL=1\n");
-
-        if (cd_media_state)
-                printf("ID_CDROM_MEDIA_STATE=%s\n", cd_media_state);
-        if (cd_media_session_next > 0)
-                printf("ID_CDROM_MEDIA_SESSION_NEXT=%u\n", cd_media_session_next);
-        if (cd_media_session_count > 0)
-                printf("ID_CDROM_MEDIA_SESSION_COUNT=%u\n", cd_media_session_count);
-        if (cd_media_session_count > 1 && cd_media_session_last_offset > 0)
-                printf("ID_CDROM_MEDIA_SESSION_LAST_OFFSET=%llu\n", cd_media_session_last_offset);
-        if (cd_media_track_count > 0)
-                printf("ID_CDROM_MEDIA_TRACK_COUNT=%u\n", cd_media_track_count);
-        if (cd_media_track_count_audio > 0)
-                printf("ID_CDROM_MEDIA_TRACK_COUNT_AUDIO=%u\n", cd_media_track_count_audio);
-        if (cd_media_track_count_data > 0)
-                printf("ID_CDROM_MEDIA_TRACK_COUNT_DATA=%u\n", cd_media_track_count_data);
-exit:
-        if (fd >= 0)
-                close(fd);
-        log_close();
-        return rc;
+        print_properties(&c);
+
+        return 0;
 }
+
+DEFINE_MAIN_FUNCTION(run);
```

Comment:
Grasping at straws, but the one difference I can think of is that the main function defined here calls some libselinux API, so when compiling -Wl,--as-needed will no longer drop the link to libselinux.so. I imagine that could very well be the initial cause of the build failure (but not the root cause).
Can you try to push a commit partially reverting the change, and keeping the old main()? Or simply redefine this macro locally without the mac_selinux_finish() call and the selinux-util.h include.

Reply:
Thank you for the investigation. Now the macro is not used in cdrom_id helper.

Reply:
it is just a hunch - haven't verified it, as I still am not able to repro locally. Fingers crossed...

Reply:
it's past the initial build - I think we got the trigger! Now we only need to figure out the root cause... I suspect the order of linker flags and libs is the culprit here (-pthread and -lselinux)

http://autopkgtest.ubuntu.com/running#pkg-systemd-upstream

```
Unpacking build-essential (12.4ubuntu1) ...
Selecting previously unselected package libsystemd-dev:s390x.
Preparing to unpack .../21-libsystemd-dev.deb ...
Unpacking libsystemd-dev:s390x (247.559.g7eda6515da.0) ...
Setting up libgomp1:s390x (8.4.0-1ubuntu1~18.04) ...
Setting up libatomic1:s390x (8.4.0-1ubuntu1~18.04) ...
Setting up libcc1-0:s390x (8.4.0-1ubuntu1~18.04) ...
Setting up linux-libc-dev:s390x (4.15.0-129.132) ...
Setting up gcc-7-base:s390x (7.5.0-3ubuntu1~18.04) ...
Setting up libsystemd-dev:s390x (247.559.g7eda6515da.0) ...
Setting up libmpc3:s390x (1.1.0-1) ...
Setting up libc-dev-bin (2.27-3ubuntu1.4) ...
Setting up libc6-dev:s390x (2.27-3ubuntu1.4) ...
Setting up libitm1:s390x (8.4.0-1ubuntu1~18.04) ...
Setting up libisl19:s390x (0.19-1) ...
Setting up libasan4:s390x (7.5.0-3ubuntu1~18.04) ...
Setting up libubsan0:s390x (7.5.0-3ubuntu1~18.04) ...
Setting up libgcc-7-dev:s390x (7.5.0-3ubuntu1~18.04) ...
Setting up cpp-7 (7.5.0-3ubuntu1~18.04) ...
Setting up libstdc++-7-dev:s390x (7.5.0-3ubuntu1~18.04) ...
Setting up cpp (4:7.4.0-1ubuntu2.3) ...
Setting up gcc-7 (7.5.0-3ubuntu1~18.04) ...
Setting up g++-7 (7.5.0-3ubuntu1~18.04) ...
Setting up gcc (4:7.4.0-1ubuntu2.3) ...
Setting up g++ (4:7.4.0-1ubuntu2.3) ...
update-alternatives: using /usr/bin/g++ to provide /usr/bin/c++ (c++) in auto mode
Setting up build-essential (12.4ubuntu1) ...
Setting up autopkgtest-satdep (0) ...
Processing triggers for man-db (2.8.3-2ubuntu0.1) ...
Processing triggers for libc-bin (2.27-3ubuntu1.4) ...
(Reading database ... 75081 files and directories currently installed.)
Removing autopkgtest-satdep (0) ...
autopkgtest [18:41:52]: test build-login: [-----------------------
build: OK
run: OK
autopkgtest [18:41:53]: test build-login: -----------------------]
autopkgtest [18:41:53]: test build-login:  - - - - - - - - - - results - - - - - - - - - -
build-login          PASS
autopkgtest [18:41:54]: test boot-and-services: preparing testbed
```

Reply:
I think there's no point in keeping this blocked until we find the root cause. I'd suggest to file an issue, prepare a branch with a revert that we can use to trigger the failure, and take it from there. In the meanwhile, this cleanup can go ahead indipendently.

Reply:
Opened an issue to track this: https://github.com/systemd/systemd/issues/18165

Reply:
Thanks.